File: 1027. Longest Arithmetic Subsequence/dp.js
/**
 * T: O(N^2), S: O(N^2)
 * @param {number[]} nums
 * @return {number}
 */
 var longestArithSeqLength = function(nums) {
    
    const n = nums.length;
    const dp = [...Array(n)].map(x => new Map());
    let max = 0;
    
    for(let i = 1; i < n; i++){
        for(let j = 0; j < i; j++){
            const diff = nums[i] - nums[j];
            dp[i].set(diff, 1 + (dp[j].get(diff) || 1))
            
            max = Math.max(max, dp[i].get(diff))
        }
    }
    
    return max;
    
 }

File: 1027. Longest Arithmetic Subsequence/dp.py
class Solution:
    def longestArithSeqLength(self, nums: List[int]) -> int:
        N = len(nums)
        dp = [{} for i in range(N)]
        res = 0
        
        for i in range(1, N):
            for j in range(i):
                diff = nums[i] - nums[j]
                dp[i][diff] = 1 + dp[j].get(diff, 1)
                res = max(res, dp[i][diff])
                
        return res;
        